From 234b0be3653d18e9e97fff060fb882eef97cd575 Mon Sep 17 00:00:00 2001
From: Panu Matilainen <pmatilai@redhat.com>
Date: Thu, 6 Mar 2025 13:57:09 +0200
Subject: [PATCH] Fix empty password field in passwd/group causing entry to be
 ignored

strtok() only handles non-empty tokens. Says so on the first line of
the description on the man page. Doh. So use our own argv splitting,
this is actually more handy anyhow.

Fixes: #3594
(backported from commit 75de02219100f381a84769d74eb58985975d49c5)
---
 lib/rpmug.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/lib/rpmug.c b/lib/rpmug.c
index 895fb50bf..1f99987dc 100644
--- a/lib/rpmug.c
+++ b/lib/rpmug.c
@@ -3,6 +3,7 @@
 #include <pwd.h>
 #include <grp.h>
 #include <errno.h>
+#include <rpm/argv.h>
 #include <rpm/rpmlog.h>
 #include <rpm/rpmstring.h>
 #include <rpm/rpmmacro.h>
@@ -73,12 +74,11 @@ static int lookup_field_in_file(const char *path, const char *val, int vcol, int
     while ((str = fgets(buf, sizeof(buf), f)) != NULL) {
 	int nf = vcol > rcol ? vcol : rcol;
 	const char *fields[nf + 1];
-	char *tok, *save = NULL;
 	int col = -1;
 
-	while ((tok = strtok_r(str, ":", &save)) != NULL) {
-	    fields[++col] = tok;
-	    str = NULL;
+	ARGV_t tokens = argvSplitString(str, ":", ARGV_NONE);
+	for (ARGV_const_t tok = tokens; tok && *tok; tok++) {
+	    fields[++col] = *tok;
 	    if (col >= nf)
 		break;
 	}
@@ -89,6 +89,7 @@ static int lookup_field_in_file(const char *path, const char *val, int vcol, int
 		rc = 0;
 	    }
 	}
+	argvFree(tokens);
     }
 
     fclose(f);
-- 
2.52.0

